kernel/src/process.rs

```diff
@@ -982,31 +1054,107 @@ impl<C: Chip> ProcessType for Process<'_, C> {
+                    self.try_restart(COMPLETION_FAULT);
+                } else {
+                    self.terminate(COMPLETION_FAULT);
+                    self.state.update(State::Faulted);
+                }
             }
             FaultResponse::Stop => {
                 // This looks a lot like restart, except we just leave the app
-                // how it faulted and mark it as `StoppedFaulted`. By clearing
+                // how it faulted and mark it as `Faulted`. By clearing
                 // all of the app's todo work it will not be scheduled, and
                 // clearing all of the grant regions will cause capsules to drop
                 // this app as well.
-                self.terminate();
+                self.terminate(COMPLETION_FAULT);
+                self.state.update(State::Faulted);
             }
         }
     }
 
+    /// Terminates and attempts to restart the process. The process
+    /// always terminates, but may or not restart based on kernel
+    /// policy and process state. It applies the restart policy
+    /// to the process.
+    ///
+    /// This function can be called when the process is in any state and
+    /// attempts to reset all of its state and re-initialize it so that it can
+    /// start running again.
+    ///
+    /// Restarting can fail for general reasons:
+    ///
+    /// 1. The kernel chooses not to restart the process based on the policy the
+    ///    kernel is using for restarting a specific process. For example, if a
+    ///    process has restarted a number of times in a row the kernel may
+    ///    decide to stop executing it.
+    ///
+    /// 2. Some state can no long be configured for the process. For example,
+    ///    the syscall state for the process fails to initialize.
+    ///
+    /// After `restart()` runs the process will either be queued to run its
+    /// `_start` function, or it will be terminated and unrunnable.
+    fn try_restart(&self, completion_code: u32) {
+        debug!("Trying to restart process {}", self.get_process_name());
+        // Terminate the process, freeing its state and removing any
+        // pending tasks from the scheduler's queue.
+        self.terminate(completion_code);
+
+        // If there is a kernel policy that controls restarts, it should be
+        // implemented here. For now, always restart.
+        let _res = self.restart();
+
+        // Decide what to do with res later. E.g., if we can't restart
+        // want to reclaim the process resources.
+    }
+
+    /// Stop and clear a process's state, putting it into the .
+    ///
+    /// This will end the process, but does not reset it such that it could be
+    /// restarted and run again. This function instead frees grants and any
+    /// queued tasks for this process, but leaves the debug information about
+    /// the process and other state intact.
+    fn terminate(&self, _completion_code: u32) {
```

Comment:
We added the completion code because it seemed useful from an API standpoint but I think we still need to discuss it in the larger context of process structure lifetime/reuse. Storing it in the process structure isn't helpful if the process is re-used: where do we store it?

Reply:
We could store it in the `debug` section of process memory, where it would be accessible for panic display or inspection. Since we don't have any mechanism to reclaim process memory right now, this would work OK.

Reply:
We could. But it would be worth discussing, and so we probably don't want to block 2.0 on this. The entire point was to create this mechanism so that we could later figure out exactly how to handle it. :)

Reply:
Created https://github.com/tock/tock/issues/2473 to track this, marking as resolved